crates/neon-sys/native/src/neon_threadsafe_cb.h

```diff
@@ -0,0 +1,99 @@
+#ifndef NEON_THREADSAFE_CALLBACK_H_
+#define NEON_THREADSAFE_CALLBACK_H_
+
+#include <uv.h>
+#include "neon.h"
+#include "v8.h"
+#include <mutex>
+
+namespace neon {
+
+    class ThreadSafeCallback {
+    public:
+        ThreadSafeCallback(v8::Isolate *isolate,
+                           v8::Local<v8::Value> self,
+                           v8::Local<v8::Function> callback): isolate_(isolate), close_(false)
+        {
+            async_.data = this;
+            uv_async_init(uv_default_loop(), &async_, async_callback);
+            // Save the this argument and the callback to be invoked.
+            self_.Reset(isolate, self);
+            callback_.Reset(isolate, callback);
+            // Save the context (aka realm) to be used when invoking the callback.
+            context_.Reset(isolate, isolate->GetCurrentContext());
+        }
+
+        void call(void *rust_callback, Neon_ThreadSafeCallbackHandler handler) {
+            {
+                std::lock_guard<std::mutex> lock(mutex_);
+                handlers_.push_back({ rust_callback, handler });
+            }
+            uv_async_send(&async_);
+        }
+
+        void close() {
+            // close is called when the rust struct is dropped
+            // this guarantees that it's called only once and
+            // that no other method (call) will be called after it.
+            close_ = true;
+            uv_async_send(&async_);
+        }
+
+        void async_call() {
```

Comment:
We could maybe call this `complete` and the static method could be `complete_callback`, for consistency with the Task API.